### PR TITLE
Base class for downloader middleware

### PR DESCRIPTION
**Pull request by jorenham**
As part of #2633

### CODE REVIEW

**Comment by eLRuLL** [comment]
what about making it optional for middlewares to be included in `DOWNLOADER_MIDDLEWARES`, as we could tell scrapy to look for all `DownloaderMiddleware` based classes. Priority could be a class parameter and a default could be specified too (so every middleware would have same priority).

Same idea for pipeline, extensions, etc.

**Comment by jorenham** [comment]
@eLRuLL I really like your idea. I'll get started on it.

File: scrapy/downloadermiddlewares/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class DownloaderMiddleware(object):
+    """Base class for downloader middleware.
+    """
+    def process_request(self, request, spider):
```
Comment (djunzu) [inline scrapy/downloadermiddlewares/__init__.py]: `BaseDownloaderMiddleware`?

Comment (redapple) [inline scrapy/downloadermiddlewares/__init__.py]: I wouldn't mind `DownloaderMiddleware`. `BaseSpider` for example was deprecated some time ago.

**Pull request closed without merging**